locator/tablets.hh

```diff
@@ -0,0 +1,177 @@
+
+#include "dht/token.hh"
+#include "utils/small_vector.hh"
+#include "locator/host_id.hh"
+#include "dht/i_partitioner_fwd.hh"
+#include "schema/schema_fwd.hh"
+#include "utils/chunked_vector.hh"
+#include "utils/hash.hh"
+
+#include <seastar/core/reactor.hh>
+#include <seastar/util/log.hh>
+
+#include <vector>
+
+namespace locator {
+
+extern seastar::logger tablet_logger;
+
+using token = dht::token;
+
+// Identifies tablet within the scope of a single tablet_map,
+// which has a scope of (table_id, token metadata version).
+// Different tablets of different tables can have the same tablet_id.
+// Different tablets in subsequent token metadata version can have the same tablet_id.
+// When splitting a tablet, one of the new tablets (in the new token metadata version)
+// will have the same tablet_id as the old one.
+using tablet_id = size_t;
+
+struct tablet_replica {
+    host_id host;
+    shard_id shard;
+
+    bool operator==(const tablet_replica&) const = default;
+};
+
+std::ostream& operator<<(std::ostream&, const tablet_replica&);
+
+using tablet_replica_set = utils::small_vector<tablet_replica, 3>;
+
+/// Stores information about a single tablet.
+struct tablet_info {
+    tablet_replica_set replicas;
+
+    std::optional<shard_id> get_shard(host_id host) const {
+        for (auto&& r : replicas) {
+            if (r.host == host) {
+                return r.shard;
+            }
+        }
+        return std::nullopt;
+    }
+
+    bool operator==(const tablet_info&) const = default;
+};
+
+/// Used for storing tablet state transition during topology changes.
+/// Describes transition of a single tablet.
+struct tablet_transition_info {
+    tablet_replica_set next;
+    tablet_replica pending_replica; // Optimization (next - tablet_info::replicas)
+
+    bool operator==(const tablet_transition_info&) const = default;
+};
+
+/// Stores information about tablets of a single table.
+class tablet_map {
+public:
+    using tablet_container = utils::chunked_vector<tablet_info>;
+private:
+    // The implementation assumes that _tablets.size() is a power of 2:
+    //
+    //   _tablets.size() == 1 << _log2_tablets
+    //
+    tablet_container _tablets;
+    size_t _log2_tablets; // log_2(_tablets.size())
+    std::unordered_map<tablet_id, tablet_transition_info> _transitions;
+public:
+    tablet_map();
+
+    /// Constructs a tablet map.
+    /// The actual number of allocated tablets may be higher due to rounding up to the power of two.
+    /// Check tablet_count().
+    ///
+    /// \param tablet_count The desired tablets to allocate.
+    explicit tablet_map(size_t tablet_count);
+
+    tablet_id get_tablet_id(token) const;
+    const tablet_info& get_tablet_info(tablet_id) const;
+    const tablet_transition_info* get_tablet_transition_info(tablet_id) const;
+    dht::token get_last_token(tablet_id id) const;
+    dht::token get_first_token(tablet_id id) const;
+    dht::token_range get_token_range(tablet_id id) const;
```

Comment:
why do we need both get_first, get_last, and get_token_range?
isn't the latter sufficient for all use cases?